core/client/fs/src/main/java/alluxio/client/block/stream/NaiveCachedGrpcDataReader.java

```diff
@@ -0,0 +1,263 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.stream;
+
+import alluxio.client.file.FileSystemContext;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.grpc.ReadRequest;
+import alluxio.grpc.ReadResponse;
+import alluxio.grpc.ReadResponseMarshaller;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.network.protocol.databuffer.NioDataBuffer;
+import alluxio.resource.CloseableResource;
+import alluxio.wire.WorkerNetAddress;
+import alluxio.resource.LockResource;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A gRPC data reader that responsible for reading from a specific block.
+ *
+ * The current implementation cached the block data from the beginning to
+ * the largest index being read.
+ *
+ * It follows GrpcDataReader protocol and takes strong assumption:
+ * Parallel read to the same file happens on the same time, so that read request is
+ * serialized by kernel
+ */
+@NotThreadSafe
+public final class NaiveCachedGrpcDataReader {
```

Comment:
Because the methods are different. 

DataReader's methods 
```
  DataBuffer readChunk() throws IOException;
  DataBuffer readChunkIfReady() throws IOException;
  long pos();
```
are not related to `bufferCachingGrpcDataReader`. bufferCachingGrpcDataReader doesn't have pos concept and it's readChunk() is private. The main publicly available method is readChunk(index) to read a specific chunk from this block.